### DISCUSSION TITLE
How could I add a parent to a node?

### DISCUSSION THREAD

**Discussion by stereomatchingkiss** [zeux/pugixml#628] [Q&A] [answered]
Original xml

```
<?xml version="1.0"?>
<measure>
     <layer>
         <A1>SomeValue</A1>
    </layer>
</measure>
```

The xml output I want

```
<?xml version="1.0"?>
<measure>
     <staff>
         <layer>
             <A1>SomeValue</A1>
        </layer>
    </staff>
</measure>
```

I can append <staff> as a child under measure by append_child, but how could I make every elements under measure as the child of <staff>? Thanks

**Comment by zeux** [comment, marked as answer]
append_child + append_move for all other children of `measure`; see https://pugixml.org/docs/manual.html#modify.move

Note that you will need to be careful about iterating, as append_move changes the sibling structure; you could do something like

```
pugi::xml_node staff = measure.append_child("staff");
while (measure.first_child() != staff)
    staff.append_move(measure.first_child());
```

**Reply by stereomatchingkiss** [reply]
> append_child + append_move for all other children of `measure`; see https://pugixml.org/docs/manual.html#modify.move
> 
> Note that you will need to be careful about iterating, as append_move changes the sibling structure; you could do something like
> 
> ```
> pugi::xml_node staff = measure.append_child("staff");
> while (measure.first_child() != staff)
>     staff.append_move(measure.first_child());
> ```


Thanks, it works, I have another question

Original xml

```
<?xml version="1.0"?>
<measure>
     <layer>
         <A1>SomeValue</A1>
    </layer>
    <rend xml:id="rhwp5l2" halign="left">pp</rend>
</measure>
```

```
pugi::xml_node staff = measure.append_child("staff");
while (measure.first_child() != staff)
    staff.append_move(measure.first_child());
```


First solution would make the <staff> node become the parent of layer and rend, how could I make it become parent of layer only?


```
<?xml version="1.0"?>
<measure>
     <staff>
         <layer>
             <A1>SomeValue</A1>
        </layer>
    </staff>
    <rend xml:id="rhwp5l2" halign="left">pp</rend>
</measure>
```

Tried 


```
 auto mnode = mei_mnode.node();
 if(auto const cname = string(mnode.name()); cname != "staff"){
     if(auto lnode = mnode.child("layer")){ 
         cout<<"cname != staff, cname == "<<cname<<endl;
          xml_node staff_node = mei_mnode.node().prepend_child();
          staff_node.set_name("staff");
                 
          while(lnode.parent() != staff_node)
               staff_node.prepend_move(lnode.first_child());
      }
 }

```

But this cause infinite loop

**Comment by stereomatchingkiss** [comment]
Fix the bug, just change 

`xml_node staff_node = mei_mnode.node().prepend_child();`

to

`xml_node staff_node = mei_mnode.node().append_child();`
